### PR TITLE
Emulate silhouette rendering for pick and _getConvexHullPointsForDrawable

### LINKED ISSUE #191: Improve `pick` performance
### Expected Behavior

Take less than a millisecond, or better, to pick a drawable from a given x, y position.

### Actual Behavior

On a Samsung Chromebook Plus a single pick can take 10s of milliseconds to execute. (Hard to tell actual performance cause the calls take so long it can bleed into the recorded time for the next pick operation.) The underlying `readPixels` method is also known to have a poorly defined execution time due to its behaviour.

### Steps to Reproduce

Projects like https://scratch.mit.edu/projects/130041250/ use pick to test if the Drawable is touching the mouse-pointer. Once that project has created all its clones, each clone tests if its touching the mouse-pointer, leading to frame times from **250 ms** (milliseconds) to **350 ms**.

### Operating System and Browser

All

### PR DESCRIPTION
**Pull request by mzgoddard**
### Resolves

#191 #192

Related to https://github.com/LLK/scratch-vm/issues/738

### Proposed Changes

Emulate silhouette rendering with Silhouette and EffectTransform types.

RenderWebGL, Drawable, and Skins use these types to take world and texture coordinates and query the opacity of a point on a Skin instance. Silhouette uses a texture coordinate to lookup a stored cell in its data to determine if the point on the Skin is opaque or translucent. EffectTransform transforms a texture coordinate into another texture coordinate based on a Drawable's effect values. Drawable.isTouching turns a world coordinate into a drawable coordinate and then into a texture coordinate, transforms that into another texture coordinate with EffectTransform and tests the final coordinate with `this.skin.isTouching`.

`pick` uses Drawable.isTouching in place of its past silhouette rendering and reading.

`_getConvexHullPoitnsForDrawable` uses Skin.isTouching and EffectTransform.transformPoint in place of its past silhouette rendering and reading.

#### Terms

I don't write these terms as official ones but to explain their use above.

**world coordinate** A twgl.v3 object with x and y values in scratch's common -240 to 240 x constraint and -180 to 180 y constraint.
**drawable coordinate** A twgl.v3 object with a x value between -0.5 and 0.5 and a y value between -0.5 and 0.5.
**texture coordinate** A twgl.v3 object with a x value between 0 and 1 and a y value between 0 and 1. Unbounded texture coordinates for isTouching methods return false.

### Reason for Changes

`pick` and `_getConvexHullPointsForDrawable` render drawable silhouettes with webgl and read back the pixels. Without combining the silhouette/touching renderings of RenderWebGL into one render per step, `gl.readPixels` will be a performance blocker while it busy waits for the WebGL context to complete its work and return the produced pixel info.

https://scratch.mit.edu/projects/130041250/ as an example case with usage of the above methods, starts by cloning a target 50 times. Each uses `if on edge, bounce` which builds a bounding box with a convex hull. Without a change like https://github.com/LLK/scratch-render/pull/188, each cloning causes all existing clones to rebuild their convex hull. Adding up that can cause systems like a Samsung Chromebook Plus to take up to **a second (<1fps)** for many of the last steps as it approaches 50 clones.

Each clone also checks if they are currently touching the mouse. So even after the 50 clones are no longer rebuilding their convex hull, they each use `RenderWebGL.pick` to see if they are touching the mouse pointer. The same Samsung Chromebook Plus takes anywhere from **180ms to 300ms (~2-12fps)** for each step.

Testing with this PR's change has shown the Samsung Chromebook Plus can run the same project at **~30fps** while cloning (without https://github.com/LLK/scratch-render/pull/188) and an easy **60fps** after.

### Test Coverage

Existing tests don't yet cover `pick` and `_getConvexHullPointsForDrawable`.

### Further Thoughts

This work could probably be extended in later work to also emulate the silhouette and color testing in methods like `RenderWebGL.isTouchingColor` with a new type called something like `ColorBuffer` which performs behaviour like `Silhouette` but returns color information instead of booleans. If we need to support another rendering behaviour like Multi-Sampling Anti-Aliasing (MSAA) we could do that adding support for that in ColorBuffer where the concern would probably best sit.

`_getConvexHullPointsForDrawable` still takes something like 1ms to run for the blocks in https://scratch.mit.edu/projects/130041250/ on a Samsung Chromebook Plus. Some further strategies could be looked into and probably added to Silhouette to cache `boundaryPoints` data. Possibly looping counter clockwise over every row and column finding the "left" most point that isn't empty. Another option may be to build less accurate hulls by stepping over a downscaled version of the same Silhouette data.

### CODE REVIEW

File: src/Drawable.js [hunk 1]
```diff
@@ -242,6 +247,47 @@ class Drawable {
...
+            twgl.m4.inverse(model, model);
+            this._inverseTransformDirty = false;
+        }
+
+        // Transfrom from world coordinates to Drawable coordinates.
+        const localPosition = twgl.m4.transformPoint(inverse, vec, __isTouchingPosition);
+
+        // Transform into texture coordinates.
+        localPosition[0] += 0.5;
+        // Invert the x axis.
```
Comment (cwillisf) [inline src/Drawable.js]: I would add a comment here indicating why the X axis must be inverted

File: src/EffectTransform.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+class EffectTransform {
+    /**
+     * Transform a texture coordinate to one that would be select after applying shader effects.
+     * @param {Drawable} drawable The drawable whose effects to emulate.
+     * @param {twgl.v3} vec The texture coordinate to transform.
+     * @param {?twgl.v3} dst A place to store the output coordinate.
+     * @return {twgl.v3} The coordinate after being transform by effects.
+     */
+    static transformPoint (drawable, vec, dst) {
```
Comment (cwillisf) [inline src/EffectTransform.js]: I'm not a huge fan of logically (though not quite literally) duplicating all this shader code here, but I guess it's worth it considering how much of a performance improvement we get from doing so. Still, I'd like to revisit this post-release.

File: src/RenderWebGL.js [hunk 3]
```diff
@@ -1152,63 +1141,70 @@ class RenderWebGL extends EventEmitter {
...
+        const CCW = function (A, B, C) {
+            return (((B[0] - A[0]) * (C[1] - A[1])) - ((B[1] - A[1]) * (C[0] - A[0])));
         };
-        for (let y = 0; y <= height; y++) {
+
+        // https://github.com/LLK/scratch-flash/blob/dcbeeb59d44c3be911545dfe54d
+        // 46a32404f8e69/src/scratch/ScratchCostume.as#L369-L413 Following
+        // RasterHull creation, compare and store left and right values that
+        // maintain a convex shape until that data can be passed to `hull` for
+        // further work.
```
Comment (cwillisf) [inline src/RenderWebGL.js]: I'm not sure anyone on the team understands this algorithm any more; it looks like it was copied into the older Scratch codebase from Numerical Recipes or something. Anyway, I guess this is another "let's revisit this post-release": I think we should try to make this code readable to the average Scratch open-source contributor.

Comment (mzgoddard) [reply src/RenderWebGL.js]: I commented the code below as a first step to make it more understandable.

File: src/EffectTransform.js (hunk 2 above)
Comment (mzgoddard) [reply src/EffectTransform.js]: Same.

File: src/RenderWebGL.js (hunk 3 above)
Comment (cwillisf) [reply src/RenderWebGL.js]: Awesome - thanks!

**Pull request merged**